functional_programming_tests/test/main.cpp

```diff
@@ -166,79 +183,117 @@ std_msgs::msg::UInt8MultiArray createUInt8MultiArrayMessageFromPath(
 };  // namespace utilities
 };  // namespace pathing
 
-// TODO: Set the return  to be a std::expected with the various errors ()
-// TODO: Make a function that takes the error
-std::shared_ptr<example_srvs::srv::GetPath::Response> generate_path_callback(
-    std::shared_ptr<example_srvs::srv::GetPath::Request> const request,
-    Map<unsigned char> const& occupancy_map,
-    pathing::PathingGeneratorFunctionType path_generater) {
-  auto const empty_response = []() {
-    auto response = std::make_shared<example_srvs::srv::GetPath::Response>();
-    response->success.data = false;
-    response->path = std_msgs::msg::UInt8MultiArray();
-    return response;
-  };
+namespace generate_path {
 
+/**
+ * @brief      Types of errors expected in the generate path callback function
+ */
+enum class error {
+  NO_OCCUPANCY_MAP,
+  INVALID_START_SIZE,
+  INVALID_GOAL_SIZE,
+  NO_VALID_PATH
+};
+
+/**
+ * @brief      Descriptions of the errors
+ */
+std::map<error, std::string> const error_description = {
+    {error::NO_OCCUPANCY_MAP, "The Occupancy Map is empty."},
+    {error::INVALID_START_SIZE,
+     "The start field in the request is not of size 2."},
+    {error::INVALID_GOAL_SIZE,
+     "The goal field in the request is not of size 2."},
+    {error::NO_VALID_PATH,
+     "There is no valid path between the start and goal."}};
```

Comment:
Is this considered the better way than the function? My reasoning for the function was that what if someone adds an error to the enum but doesn't add it to the map? Then the lookup with the value thats not added will be a bug, right? With the function at least theres some sort of indication on what is going wrong?

Reply:
@griswaldbrooks

Reply:
If I remember correctly, the issue is that if you add an enum value but don't update the function, you'll just get the `default`. I think if you do it and don't update the `map` you get a compile error? Let me check. It also fits in with the functional style using a map for finite values and a function for infinite values?

Reply:
https://godbolt.org/z/vfM18zzba using map
https://godbolt.org/z/81MaaWbz8 using function

ok, it's not that it doesn't compile. the map version throws, and the function version just gives you the default value